index.bs

```diff
@@ -253,18 +253,251 @@ Control Protocols {#control}
+consist of at least 16 ASCII characters.
+
+
+When the receiver receives the presentation-start-request, it should send back a
+presentation-start-response message after either the presentation URL has been
+feteched and loaded or the receiver has failed to do so. If it has failed, it
+must respond with the appropriate result (such as invalid-url or timeout).  If
+it has succeeded, it must reply with a success result.  Additionally, the
+response must include the following:
+
+- connection-id: An ID that both endpoints can use to send connection messages
+to each other.  It is chosen by the receiver for ease of implementation (if the
+message receiver chooses the connection-id, it may keep the ID unique across
+connections, thus making message demuxing/routing easier).
+
+<!-- TODO: Add optional HTTP response code to the response? -->
+
+To send a presentation message, the controller or receiver may send a
+presentation-connection-message with the following values:
+
+- connection-id: The ID from the presentation-start-response or
+presentation-connection-open-response messages.
+
+- message: the presentation message data.  
+
+
+To terminate a presentation, the controller may send a
+presentation-termination-request message with the following values:
+
+- presentation-id: The ID of the presentation to terminate.
+
+- reason: The reason the presentation is being terminated.
+
+
+When a receiving endpoint receives a presentation-termination-request, it should
+send back a presentation-termination-response message to the requesting
+endpoint.  It may also notify other controllers about the termination by sending
+a presentation-termination-event message.  And it can send the same message if
+it terminates a presentation without a request from a controller to do so. This
+message contains the following values:
+
+- presentation-id: The ID of the presentation that was terminated.
+
+- reason: The reason the presentation was terminated.
+
+<!-- TODO: Split up reason into reason and whether it was triggered by the user
+or not? -->
+
+
+To accept incoming connections requests from controller, a receiver
+must receive and process the presentation-connection-open-request
+message which contains the following values: 
+
+- presentation-id: The ID of the presentation to connect to.
+
+
+The receiver should, upon receipt of a
+presentation-connection-open-request message, send back a
+presentation-connection-open-response message which contains the
+following values:
+
+- result: a code indicating success or failure, and the reason for the failure
+
+- connection-id: An ID that both endpoints can use to send connection messages
+to each other.  It is chosen by the receiver for ease of implementation (if the
+message receiver chooses the connection-id, it may keep the ID unique across
+connections, thus making message demuxing/routing easier).
+
+
+A controller may terminate a connection without terminating the presentation by
```

Comment:
I've been wondering if, for both close and terminate, we should exclusively use an event.  Is there any reason for the controller to wait for a close-response/termination-response?  Is there anything it should do with an error in one of those responses?

Reply:
The receiver could refuse to terminate for some reason (maybe it decides the controller isn't allowed to terminate that presentation).  The controller can't assume it happened until it receives the response.  The presentation API doesn't allow for a close() or terminate() method to fail asynchronously, though.  So if we wanted to allow the JS to do something, like indicate to the user that close/terminate failed, it wouldn't have a way of knowing.

Reply:
Yeah.  close/terminate always creates a state change on the controller side.  Since the controller can't assume it can communicate with the presentation (or the device it's on) after these are called, there's no way to communicate the outcome back to the JS.

Reply:
Couldn't we change the close() method to return a Promise to resolve when the close is acked?